Change Scheduler ComponentConfig fields to nilable pointers

damemi opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
/kind api-change

**What this PR does / why we need it**:
This changes some basic Scheduler ComponentConfig fields to pointers, so that we can check nil (unset) against defaults. It also adds nil checks and updates defaulting logic for these fields. See https://github.com/kubernetes/kubernetes/issues/83575 for more

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Scheduler ComponentConfig fields are now pointers
```

damemi commented:
/cc @ahg-g 
Could you look at this and let me know if it's the right direction? I could continue this for the rest of the fields too

ahg-g commented:
> /cc @ahg-g
> Could you look at this and let me know if it's the right direction? I could continue this for the rest of the fields too

yes, it's the right direction.

/cc @liggitt

fejta commented:
/uncc

ahg-g commented on pkg/scheduler/apis/config/v1alpha1/defaults.go:
```diff
@@ -55,22 +57,37 @@ func SetDefaults_KubeSchedulerConfiguration(obj *kubeschedulerconfigv1alpha1.Kub
+	}
+
+	if obj.PercentageOfNodesToScore == nil {
+		percentageOfNodesToScore := int32(50)
```
please move the const in [api/types.go](https://github.com/kubernetes/kubernetes/blob/b8b7c376edb62ce8be42cef3728112c5c71d9fa4/pkg/scheduler/api/types.go#L29) to apis/config/types.go and use it here.

damemi replied:
done

ahg-g reviewed (commented):
Can you change the [Weight](https://github.com/kubernetes/kubernetes/pull/83619/files#diff-b483008d22340548275bf45d61561d1dR210) to a pointer as well, this is also optional. The validation for this field is: if set, should be larger than 0.

damemi commented:
Sorry, @ahg-g your first link doesn't go to a specific line so I'm not sure which `Weight` you're referring to. Do you mean the [`HardPodAffinitySymmetricWeight`](https://github.com/kubernetes/kubernetes/pull/83619/files#diff-b483008d22340548275bf45d61561d1dR54)`?

ahg-g commented:
> Sorry, @ahg-g your first link doesn't go to a specific line so I'm not sure which `Weight` you're referring to. Do you mean the [`HardPodAffinitySymmetricWeight`](https://github.com/kubernetes/kubernetes/pull/83619/files#diff-b483008d22340548275bf45d61561d1dR54)`?

no, I meant this one: https://github.com/kubernetes/kubernetes/blob/f7992c7a214cc9ffbf43a4d340c1c3e913f667e3/pkg/scheduler/apis/config/types.go#L210

ahg-g commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -238,9 +238,13 @@ func (o *Options) Config() (*schedulerappconfig.Config, error) {
 	eventBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{Interface: eventClient.EventsV1beta1().Events("")})
-	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, c.ComponentConfig.SchedulerName)
+	reportingController := ""
+	if c.ComponentConfig.SchedulerName != nil {
```
instead of defaulting to "", we should error fatal if schedulerName is not set or empty string.

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -172,9 +172,13 @@ func NewFramework(r Registry, plugins *config.Plugins, args []config.PluginConfi
+		f.pluginNameToWeightMap[name] = weight
 		if f.pluginNameToWeightMap[name] == 0 {
 			f.pluginNameToWeightMap[name] = 1
 		}
```
can be replaced with:
```
if pg[name].Weight != nil && int(*pg[name].Weight) != 0 {
  f.pluginNameToWeightMap[name] = int(*pg[name].Weight)
} else {
  f.pluginNameToWeightMap[name] = 1
}
```

ahg-g commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -151,9 +151,14 @@ func TestSchedulerCreationFromConfigMap(t *testing.T) {
 					{Name: "VolumeZone"},
 				},
 				"ScorePlugin": {
+<<<<<<< HEAD
```
resolve the diffs.

damemi commented:
/label api-review

ahg-g commented:
/approve

liggitt commented on cmd/kube-scheduler/app/options/insecure_serving.go:
```diff
@@ -53,17 +53,17 @@ func (o *CombinedInsecureServingOptions) AddFlags(fs *pflag.FlagSet) {
 
 func (o *CombinedInsecureServingOptions) applyTo(c *schedulerappconfig.Config, componentConfig *kubeschedulerconfig.KubeSchedulerConfiguration) error {
-	if err := updateAddressFromDeprecatedInsecureServingOptions(&componentConfig.HealthzBindAddress, o.Healthz); err != nil {
+	if err := updateAddressFromDeprecatedInsecureServingOptions(componentConfig.HealthzBindAddress, o.Healthz); err != nil {
```
Most of these methods assume that the passed-in pointer is non-nil. Using a value that can now be nil adds the risk of a NPE, for little apparent benefit.

liggitt commented:
> This changes some basic Scheduler ComponentConfig fields to pointers, so that we can check nil (unset) against defaults. It also adds nil checks and updates defaulting logic for these fields.

Would it be safer (and less disruptive to the rest of the scheduling code) to keep the internal type fields as non-pointers, and limit the pointers to the serialized config types?

damemi commented:
> Would it be safer (and less disruptive to the rest of the scheduling code) to keep the internal type fields as non-pointers, and limit the pointers to the serialized config types?

I don't oppose this, @ahg-g would that work for us?

ahg-g commented:
> > Would it be safer (and less disruptive to the rest of the scheduling code) to keep the internal type fields as non-pointers, and limit the pointers to the serialized config types?
> 
> I don't oppose this, @ahg-g would that work for us?

Me neither, but in general, I thought that [the defaulting logic](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/v1alpha1/defaults.go) is the one responsible for properly initializing those pointers if not set, and so we should generally be ok dereferncing them.

damemi commented:
@ahg-g does that defaulting logic apply to the same types that @liggitt mentioned though? It looks like in those places we are dereferencing from a `k8s.io/kubernetes/pkg/scheduler/apis/config` scheduler config, but the defaults get set on a config from `k8s.io/kube-scheduler/config/v1alpha1`

ahg-g commented:
> @ahg-g does that defaulting logic apply to the same types that @liggitt mentioned though? It looks like in those places we are dereferencing from a `k8s.io/kubernetes/pkg/scheduler/apis/config` scheduler config, but the defaults get set on a config from `k8s.io/kube-scheduler/config/v1alpha1`

I am not sure I understand, what I meant is that before the scheduler gets to examine the config (and dereference its fields), the defaulting logic is executed on the config, and it is supposed to properly initialize the fields if there is a default for that field. Fields that don't have a default value means that a nil value is something valid and the scheduler should handle it properly.

liggitt commented:
> I am not sure I understand, what I meant is that before the scheduler gets to examine the config (and dereference its fields), the defaulting logic is executed on the config, and it is supposed to properly initialize the fields if there is a default for that field. Fields that don't have a default value means that a nil value is something valid and the scheduler should handle it properly.

Distinguishing between nil and "" is most useful when "" is a valid value, and you want to be able to default to something other than "".

For the first few fields I looked at, there was no clear benefit in making all users of the internal types check for nil before dereferencing in order to avoid nil pointer panics.

For fields which cannot be nil, making the internal field a non-pointer ensures that if defaulting does not populate the field, *string becomes "" internally and cannot cause panics.

liggitt commented:
If this was a new config file, I would be less concerned, but there are many code paths here and I'm not confident we have exercised all of them with all possible values in order to flush out assumptions and potential nil panics.

ahg-g commented:
Ok, so you are suggesting not to move forward with this. I am ok with that, my understanding was that graduating the scheduler's component config to beta is blocked on #78109 (see [here](https://github.com/kubernetes/enhancements/issues/785#issuecomment-531669546)), and I wanted to unblock us by actually doing the work. Do you have any concerns graduating the scheduler's component config to beta in its current state?

liggitt commented:
> Ok, so you are suggesting not to move forward with this.

It's reasonable to change the fields in the external types to pointers and switch defaulting to only apply to nil values (not "" or 0), and that would be good to do before graduating to beta.

ahg-g commented:
> > Ok, so you are suggesting not to move forward with this.
> 
> It's reasonable to change the fields in the external types to pointers and switch defaulting to only apply to nil values (not "" or 0), and that would be good to do before graduating to beta.

Sounds good.

ingvagabund commented on cmd/kube-scheduler/app/options/insecure_serving_test.go:
```diff
@@ -42,8 +48,8 @@ func TestOptions_ApplyTo(t *testing.T) {
 				ComponentConfig: kubeschedulerconfig.KubeSchedulerConfiguration{
-					HealthzBindAddress: "1.2.3.4:1234",
-					MetricsBindAddress: "1.2.3.4:1234",
+					HealthzBindAddress: servingAddress("1.2.3.4:1234"),
```
You might use `k8s.io/utils/pointer.StringPtr` [1] instead.

[1] https://github.com/kubernetes/utils/blob/master/pointer/pointer.go#L74

ingvagabund commented on cmd/kube-scheduler/app/options/options_test.go:
```diff
@@ -528,3 +532,19 @@ pluginConfig:
 	}
 }
+
+func stringPtr(s string) *string {
```
https://github.com/kubernetes/utils/blob/master/pointer/pointer.go has all the functions bellow already defined.

damemi replied:
Oh great! Wasn't aware of that package, thanks

ahg-g commented:
Hi Mike, are you making progress on this one? Jordan's idea is to make the external types pointer based (i.e., the ones define in staging/src/k8s.io/kube-scheduler/config/v1alpha1/types.go), but the internal ones not (i.e., the ones defined in pkg/scheduler/apis/config/types.go ).

damemi commented:
/retest

damemi commented:
@ahg-g updated with @liggitt's feedback, if this can be approved I'll squash down

ahg-g commented:
/assign @alculquicondor

liggitt commented on pkg/scheduler/apis/config/v1alpha1/defaults.go:
```diff
@@ -55,22 +58,37 @@ func SetDefaults_KubeSchedulerConfiguration(obj *kubeschedulerconfigv1alpha1.Kub
 		}
-		obj.HealthzBindAddress = net.JoinHostPort(host, port)
-	} else {
-		obj.HealthzBindAddress = net.JoinHostPort("0.0.0.0", strconv.Itoa(ports.InsecureSchedulerPort))
```
this case got dropped (existing value is non-nil, but does not include port)

damemi replied:
Now that you mention it, it looks like this didn't quite cover that case originally either, since it defaults the address to `0.0.0.0` even if it's just missing the port, so I'll fix these up to use the address that's set and default to `0.0.0.0` only if that's empty

damemi replied:
In my changes though, I did cover these `else` cases by declaring the `defaultBindAddress` above both, and setting it before the error checks (as opposed to after)

liggitt replied:
> In my changes though, I did cover these `else` cases by declaring the `defaultBindAddress` above both (https://github.com/kubernetes/kubernetes/pull/83619/files/b019c2ec6ba787ac1c24ce291fb8490420d951af#diff-9d72b6493a5482d8a2829af1390c5733R61), and setting it before the error checks (as opposed to after)

I don't think that's accurate. You only set it in the nil case, when the value is completely absent from the config. When just the address (`0.0.0.0`) or just the port (`:1234`) is set, the nil path would not be exercised.

damemi replied:
Actually, it seems that `net.SplitHostPort` doesn't even return the host if the port is missing ([playground example](https://play.golang.org/p/j-F_PGIoggO)), so it becomes more difficult here to properly implement the case you described. If the existing value is `1.2.3.4` (no port), then `SplitHostPort` only returns an error `missing port in address`. We could do parsing (regex?) on our own to determine if the original value is a valid host, then append the port, while we would also have to string-compare the error message itself, while making sure there are no other errors in it. @liggitt what do you think?

liggitt replied:
would this work: if splithostport returns an err, assume the value does not have a port, call call net.ParseIP on the value to make sure it is a valid bind address, and if that returns non-nil, append the default port?

damemi replied:
@liggitt you are right, my original changes would only work in the nil case. I've added the suggestion you make the check the output from net.ParseIP, and I think the combination of these has enabled us to check each edge case. For confirmation, I refactored the unit test for these defaults to use table tests and added these cases. Please take another look, and thank you for your feedback so far

damemi replied:
(changes pushed to new commit)

liggitt commented on pkg/scheduler/apis/config/v1alpha1/defaults.go:
```diff
@@ -55,22 +58,37 @@ func SetDefaults_KubeSchedulerConfiguration(obj *kubeschedulerconfigv1alpha1.Kub
 		}
-		obj.MetricsBindAddress = net.JoinHostPort(host, port)
-	} else {
-		obj.MetricsBindAddress = net.JoinHostPort("0.0.0.0", strconv.Itoa(ports.InsecureSchedulerPort))
```
this case got dropped as well, an existing value without a port should get the default port appended

liggitt commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -651,7 +652,8 @@ func (f *framework) ListPlugins() map[string][]config.Plugin {
 				// Weights apply only to score plugins.
-				p.Weight = int32(f.pluginNameToWeightMap[name])
+				pluginWeight := int32(f.pluginNameToWeightMap[name])
+				p.Weight = pluginWeight
```
this looks like a no-op diff now

liggitt commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -184,10 +184,11 @@ func NewFramework(r Registry, plugins *config.Plugins, args []config.PluginConfi
 		// when configured.
-		f.pluginNameToWeightMap[name] = int(pg[name].Weight)
-		if f.pluginNameToWeightMap[name] == 0 {
+		if int(pg[name].Weight) != 0 {
```
this also looks like a no-op diff now

liggitt replied:
can revert this diff, it is the same as lines 189-192 above

damemi replied:
sorry, missed that

liggitt commented:
this looks a lot cleaner/safer. a couple comments on defaulting behavior that changed (needs covering unit tests for those) and a couple spurious diffs.

liggitt commented on pkg/scheduler/apis/config/v1alpha1/defaults.go:
```diff
@@ -55,22 +58,62 @@ func SetDefaults_KubeSchedulerConfiguration(obj *kubeschedulerconfigv1alpha1.Kub
+				hostPort := net.JoinHostPort(*obj.HealthzBindAddress, strconv.Itoa(ports.InsecureSchedulerPort))
+				obj.HealthzBindAddress = &hostPort
+			} else {
+				obj.HealthzBindAddress = &defaultBindAddress
```
this is good for backwards compatibility, but add a TODO that in v1beta1 we'd just want to let this error, rather than stomping it with a default ip+port

liggitt commented on pkg/scheduler/apis/config/v1alpha1/defaults.go:
```diff
@@ -55,22 +58,62 @@ func SetDefaults_KubeSchedulerConfiguration(obj *kubeschedulerconfigv1alpha1.Kub
+				hostPort := net.JoinHostPort(*obj.MetricsBindAddress, strconv.Itoa(ports.InsecureSchedulerPort))
+				obj.MetricsBindAddress = &hostPort
+			} else {
+				obj.MetricsBindAddress = &defaultBindAddress
```
same TODO about letting this case error in v1beta1

damemi commented:
@liggitt force pushed with your new feedback

liggitt commented:
compile error in unit test, and needs squash, but lgtm otherwise

damemi committed:
Change scheduler ComponentConfig fields to nilable

As part of graduating the scheduler's component config to beta, we require configurable fields to be nilable pointers (see https://github.com/kubernetes/kubernetes/issues/78109). This enables the ability to distinguish between default and unset values. We are only applying this change to external types, and reacting in our defaulting logic. This also reverts existing internal component config fields which were pointers to be non-pointers, for consistency.

liggitt commented:
/approve

ahg-g commented:
/retest

ahg-g reviewed (commented):
Thanks, can you change the existing pointer fields in the internal type to non-pointers: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/types.go#L91-L101

damemi commented:
@ahg-g I pushed a new commit reverting those types you linked to, ptal

ahg-g commented:
Thanks, please squash.

ahg-g commented:
/lgtm

damemi commented on pkg/scheduler/apis/config/v1alpha1/defaults.go:
```diff
@@ -35,12 +36,14 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
-	if obj.HardPodAffinitySymmetricWeight == 0 {
-		obj.HardPodAffinitySymmetricWeight = api.DefaultHardPodAffinitySymmetricWeight
+	if obj.HardPodAffinitySymmetricWeight == nil {
+		defaultHardPodAffinitySymmetricWeight := api.DefaultHardPodAffinitySymmetricWeight
```
Personally I think this makes the file less readable, but sure. `default` is a reserved keyword in go, so switched to `val`.

alculquicondor replied:
I actually think that it makes the purpose more obvious: that this is a local variable.

damemi replied:
Good point, that's fair enough for me 👍

damemi commented:
@alculquicondor can you re-lgtm?

alculquicondor commented:
/lgtm

alculquicondor commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -236,6 +236,9 @@ func (o *Options) Config() (*schedulerappconfig.Config, error) {
 	}
 
 	coreBroadcaster := record.NewBroadcaster()
+	if c.ComponentConfig.SchedulerName == "" {
```
wait, can you revert this? I'm adding the checks in ApplyTo #84815

damemi replied:
Yup, removed, can you re-lgtm?

alculquicondor commented:
/hold

alculquicondor commented:
Thanks!

/lgtm

/hold cancel

damemi commented:
/retest
